agent/dns.go

```diff
@@ -1483,10 +1369,284 @@ func findWeight(node structs.CheckServiceNode) int {
+
+	if serviceNode.Service.Address != "" {
+		addr = serviceNode.Service.Address
+	}
+
+	// If the service address is a CNAME for the service we are looking
+	// for then use the node address.
+	if dnsQuery == strings.TrimSuffix(addr, ".")+"." {
+		addr = serviceNode.Node.Address
+	}
+
+	return addr
+}
+
+func (d *DNSServer) encodeIPAsFqdn(dc, ip string) string {
+	addr := net.ParseIP(ip)
+	ipv4 := addr.To4()
+	if ipv4 != nil {
+		ipStr := hex.EncodeToString(addr)
+		return fmt.Sprintf("%s.addr.%s.%s", ipStr[len(ipStr)-(net.IPv4len*2):], dc, d.domain)
+	} else {
+		return fmt.Sprintf("%s.addr.%s.%s", hex.EncodeToString(addr), dc, d.domain)
+	}
+}
+
+func makeARecord(qType uint16, addr string, ttl time.Duration) dns.RR {
+	ip := net.ParseIP(addr)
+
+	var ipRecord dns.RR
+	ipv4 := ip.To4()
+	if ipv4 != nil {
+		if qType == dns.TypeSRV || qType == dns.TypeA || qType == dns.TypeANY || qType == dns.TypeNS || qType == dns.TypeTXT {
+			ipRecord = &dns.A{
+				Hdr: dns.RR_Header{
+					Rrtype: dns.TypeA,
+					Class:  dns.ClassINET,
+					Ttl:    uint32(ttl / time.Second),
+				},
+				A: ipv4,
+			}
+		}
+	} else if qType == dns.TypeSRV || qType == dns.TypeAAAA || qType == dns.TypeANY || qType == dns.TypeNS || qType == dns.TypeTXT {
+		ipRecord = &dns.AAAA{
+			Hdr: dns.RR_Header{
+				Rrtype: dns.TypeAAAA,
+				Class:  dns.ClassINET,
+				Ttl:    uint32(ttl / time.Second),
+			},
+			AAAA: ip,
+		}
+	}
+	return ipRecord
+}
+
+// Craft dns records for a node
+// In case of an SRV query the answer will be a IN SRV and additional data will store an IN A to the node IP
+// Otherwise it will return a IN A record
+func (d *DNSServer) makeRecordFromNode(dc string, node *structs.Node, qType uint16, qName string, ttl time.Duration, maxRecursionLevel int) []dns.RR {
+	addr := d.agent.TranslateAddress(dc, node.Address, node.TaggedAddresses)
+
+	var res []dns.RR
+
+	if net.ParseIP(addr) == nil {
+		res = append(res, &dns.CNAME{
+			Hdr: dns.RR_Header{
+				Name:   qName,
+				Rrtype: dns.TypeCNAME,
+				Class:  dns.ClassINET,
+				Ttl:    uint32(ttl / time.Second),
+			},
+			Target: dns.Fqdn(node.Address),
+		})
+
+		res = append(res,
+			d.resolveCNAME(d.config.Load().(*dnsConfig), dns.Fqdn(node.Address), maxRecursionLevel)...,
+		)
+
+		return res
+	}
+
+	ipRecord := makeARecord(qType, addr, ttl)
+	if ipRecord == nil {
+		return nil
+	}
+
+	ipRecord.Header().Name = qName
+	return []dns.RR{ipRecord}
+}
+
+// Craft dns records for a service
+// In case of an SRV query the answer will be a IN SRV and additional data will store an IN A to the node IP
+// Otherwise it will return a IN A record
+func (d *DNSServer) makeRecordFromServiceNode(dc string, serviceNode structs.CheckServiceNode, req *dns.Msg, ttl time.Duration) ([]dns.RR, []dns.RR) {
+	q := req.Question[0]
+	addr := d.agent.TranslateAddress(dc, serviceNode.Node.Address, serviceNode.Node.TaggedAddresses)
+	ipRecord := makeARecord(q.Qtype, addr, ttl)
```

Comment:
If the addr is a hostname then wont this prevent returning anything for the service node?

Reply:
@mkeeler As fqdn case is already addressed what type of hostname would you expect here?

Reply:
@mkeeler Could you please clarify this for me?

Reply:
> If the addr is a hostname then wont this prevent returning anything for the service node?

Hello @mkeeler ... @tionebsalocin nor @Aestek nor me really understand what you mean... can you clarify please ? (we would like this to get merged before working on IPv6 support)

The tests are now all passing but ARM (recently added)

Thank you!

Reply:
@pierresouchay `d.agent.TranslateAddress` will return one of the nodes addresses. What if the WAN address that is within the TaggedAddresses is a hostname. In that case makeARecord could at worst segfault trying to convert the nil ip addr to v4. If it doesn't segfault then I think it will end up with an invalid AAAA record being created.

Reply:
@mkeeler The check was done already on the translated addr. We cleaned this part a bit to only translate and parse ip addresses once.

Reply:
@Aestek the code for the PR seems to have disappeared. Once its back up I will give this a thorough review again and hopefully get a better grasp on how the hostname is being accounted for.

Reply:
Woops sorry I failed at git, fixed now